src/System.Runtime.Extensions/src/System/Net/WebUtility.cs

```diff
@@ -357,6 +358,93 @@ private static byte[] UrlEncode(byte[] bytes, int offset, int count)
             return expandedBytes;
         }
 
+        private static unsafe string UrlEncode(string originalString, char* s, int length)
+        {
+            int numSpaces = 0;
+            int numUnsafe = 0;
+            int numSafe = 0;
+            for (int i = 0; i < length; i++)
+            {
+                char c = s[i];
+                if (c == ' ')
+                {
+                    numSpaces++;
+                }
+                else if (!IsUrlSafeChar(c))
+                {
+                    numUnsafe += GetByteCount(s, i, length);
+                }
+                else
+                {
+                    numSafe++;
+                }
+            }
+
+            if (numSpaces == 0 && numUnsafe == 0)
+            {
+                // Nothing to expand
+                return originalString;
+            }
+
+            byte[] originalBytes = new byte[numSafe + numUnsafe + numSpaces];
+            byte[] newBytes = new byte[originalBytes.Length + numUnsafe * 2];
+
+            fixed (byte* originalBytesPointer = originalBytes)
+            fixed (byte* newBytesPointer = newBytes)
+            {
+                Encoding.UTF8.GetBytes(s, length, originalBytesPointer, originalBytes.Length);
```

Comment:
Note: we don't use the `char*` parameter past this point, so we're unnecessarily pinning the string while we're making a pass over `originalBytes`. You may want to refactor a bit to avoid that.